autoextract_poet/retry.py

```diff
@@ -0,0 +1,162 @@
+"""
+AutoExtract additional retrying logic to the one
+provided by tha AutoExtract client.
+
+The general policy is:
+
+* No retry on deterministic errors (e.g. Wrong API key, malformed query, etc)
+* Retry when there are chances to obtain a different response
+  (e.g. after query timeout or in the presence of HTTP 500 from target site)
+* Retry after the suggested time for the "domain is occupied ..." case
+
+See https://doc.scrapinghub.com/autoextract.html#errors
+
+"""
+import logging
+import re
+from typing import Optional
+
+from tenacity import (
+    RetryCallState,
+    after_log,
+    before_sleep_log,
+    retry,
+    retry_if_exception,
+    stop_after_delay,
+    wait_fixed,
+    wait_random,
+    wait_random_exponential,
+)
+from tenacity.stop import stop_base, stop_never
+from tenacity.wait import wait_base
+
+from autoextract_poet.exceptions import QueryLevelError
+
+logger = logging.getLogger(__name__)
+
+_DOMAIN_OCCUPIED_RE = re.compile(
+    r".*domain .* is occupied, please retry in (.*) seconds.*", re.IGNORECASE
+)
+
+
+def _domain_occupied(msg: Optional[str]) -> Optional[float]:
```

Comment:
That's awful we need to write such code in a client library; API should have exposed this in a structured format. That's good to ahve support for it in the library now, but I hope we'll be able to remove it in future.